### PR TITLE
feat: snappy framing format (decompression)

### LINKED ISSUE #224: [Snappy] Implement Snappy framing format in Elixir
_Needs some thinking_

Since the framing format is fairly simple, it's worth considering implementing [the framing format](https://github.com/google/snappy/blob/main/framing_format.txt) ourselves in Elixir (using `:snappyer` for the underlying block compression) to remove the complexity of interfacing with Rust code. This should simplify the implementation of a true streaming de/compression (as of now, it needs the whole stream in memory to work).

### PR DESCRIPTION
**Pull request by MegaRedHand**
Closes #224
Depends on #713

This PR continues the work done in #543. What's missing is implementing Snappy streaming compression, and running some benchmarks comparing it to the `Snappy` NIF.

### CODE REVIEW

File: test/unit/snappyex_test.exs [hunk 1]
```diff
@@ -0,0 +1,108 @@
+defmodule Unit.SnappyExTest do
+  use ExUnit.Case
+  use ExUnitProperties
+  doctest SnappyEx
+
+  @empty_stream <<0xFF, 6::little-size(24)>> <> "sNaPpY"
+
+  def assert_snappy_decompress(compressed, uncompressed) do
```
Comment (Arkenan) [inline test/unit/snappyex_test.exs:8]: defp?

File: test/unit/snappyex_test.exs [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+    assert_snappy_decompress(compressed_payload, expected)
+
+    # Compressed chunks
+    msg = "0011FF060000734E61507059000A0000B3A056EA1100003E0100"
+    "00" <> "11" <> compressed_payload = msg
+    expected = Base.decode16!("0000000000000000000000000000000000")
+
+    assert_snappy_decompress(compressed_payload, expected)
+  end
```
Comment (Arkenan) [inline test/unit/snappyex_test.exs:66]: These should be 3 separate tests.

File: test/unit/snappyex_test.exs [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    stream = Enum.join([@empty_stream, header, checksum, compressed_data])
+
+    assert SnappyEx.decompress(stream) == {:ok, data}
+  end
+
+  test "decompress GetMetadata responses" do
+    # Uncompressed chunks
+    msg = "0011FF060000734E6150705901150000F1D17CFF0008000000000000FFFFFFFFFFFFFFFF0F"
+    # status <> length <> ...
+    "00" <> "11" <> compressed_payload = msg
```
Comment (Arkenan) [inline test/unit/snappyex_test.exs:49]: Why do we have "0011" in the message if we remove it afterwards?

File: test/unit/snappyex_test.exs [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+  end
+
+  test "decompress Ping responses" do
+    for {msg, expected} <- [
+          {"0008FF060000734E61507059010C0000B18525A04300000000000000", "4300000000000000"},
+          {"0008FF060000734E61507059010C00000175DE410100000000000000", "0100000000000000"},
+          {"0008FF060000734E61507059010C0000EAB2043E0500000000000000", "0500000000000000"},
+          {"0008FF060000734E61507059010C0000290398070000000000000000", "0000000000000000"}
+        ] do
+      "00" <> "08" <> compressed_payload = msg
```
Comment (Arkenan) [inline test/unit/snappyex_test.exs:75]: Same here.

File: test/unit/snappyex_test.exs [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+      expected = Base.decode16!(expected)
+
+      assert_snappy_decompress(compressed_payload, expected)
+    end
+
+    # Error response
+    msg =
+      "011CFF060000734E6150705900220000EF99F84B1C6C4661696C656420746F20756E636F6D7072657373206D657373616765"
+
+    "01" <> "1C" <> compressed_payload = msg
```
Comment (Arkenan) [inline test/unit/snappyex_test.exs:85]: Same here. If this is a non-snappy related header (e.g. gossip identifier) we may either not put it in the test or have a function that removes the first four bytes in every test, but we're not using any of those for testing.

**Review by Arkenan — APPROVED** [review]
LGTM, added some nit comments on testing.

File: test/unit/snappyex_test.exs (hunk 3 above)
Comment (MegaRedHand) [reply test/unit/snappyex_test.exs:48]: The test was copied from the first snappy tests we had. This was intended as an example of manually decoding a message from a peer. Do you think we should move it somewhere else? Or should I just remove it?

Comment (MegaRedHand) [reply test/unit/snappyex_test.exs:48]: I just removed it. We can get new samples from peers anyways.

**Review by Arkenan — APPROVED** [review]
LGTM

**Pull request merged**